### PR TITLE
Quantization with torch FX [part 2]

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This is the second PR in the process of adding PyTorch based quantization features to `optimum`. This PR continues what was started in the previous PR #216, by adding:

- 3 quantization configuration classes: those classes are directly mapping what is being done in PyTorch with `torch.ao.quantization.QConfig` and the `qconfig_dict`, but they are serializable, and thus can be uploaded to the HuggingFace Hub.
  - `QConfigUnit`: this is the `optimum` equivalent of `torch.ao.quantization.Observer` and `torch.ao.quantization.FakeQuantize`, it specifies how quantization values will be calibrated.
  - `QConfig`: it directly maps to `torch.ao.quantization.QConfig`, and does nothing more, except that it uses `QConfigUnit`s instead of `torch.ao.quantization.Observer`s or `torch.ao.quantization.FakeQuantize`s.
  - `QuantizationConfig`: this is the class that specifies how a model should be quantized, this is the `optimum` equivalent of the `qconfig_dict` in PyTorch. This is this class that can be uploaded to the Hub.
 - 2 calibration methods: one for post training static quantization, and the other for quantization aware training.
 - The `quantize` function which takes a model as input, and returns its quantized counterpart. 

These functions are still agnostic to the `optimum` quantization configuration API, and work directly with the PyTorch API. This enable users to use this if needed, and the plan is to compose all of theme together in a class called `FXQuantizer` or something like that in next PRs.

### CODE REVIEW

**Comment by fxmarty** [comment]
Will be really useful to try out graph modifications on top of quantization I think!

By the way, something is probably broken on the live PR doc generation, the changes do not appear. Maybe due to a change here https://github.com/huggingface/optimum/pull/299/files#diff-3f9cec68c26c8d0918fecf5f80cf447e52ae3cfbf154018e0eab35a8dfc0e18c ?

File: optimum/fx/quantization/configuration.py [hunk 1]
```diff
@@ -0,0 +1,862 @@
...
+        quant_max (`int`, *optional*):
+            The maximum quantization value.
+        calibration_method (`CalibrationMethod`, defaults to `CalibrationMethod.MinMax`):
+            The calibration method to use.
+        averaging_constant (`float`, defaults to 0.01):
+            Averaging constant for min/max, used when `calibration_method=CalibrationMethod.MovingAverage`.
+        bins (`int`, defaults to 2048):
+            The number of bins to use, used when `calibration_method=CalibrationMethod.Histogram`.
+        upsample_rate (`float`, defaults to 128):
+            The factor by which the histograms are upsampled, used `calibration_method=CalibrationMethod.Histogram`.
```
Comment (regisss) [inline optimum/fx/quantization/configuration.py:91]:
```suggestion
            The factor by which the histograms are upsampled, used when `calibration_method=CalibrationMethod.Histogram`.
```

File: tests/fx/quantization/test_configuration.py [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+                    self.assertTrue(
+                        torch.allclose(t1, t2),
+                        f"The example do not match between the PyTorch ref and the restored to PyTorch one, max diff: {(t1 - t2).abs().max()}, classes_and_kwargs = {classes_and_kwargs[idx]}",
+                    )
+
+        # From observer back to observer
+        test_fn(lambda class_, kwargs: class_(**kwargs), lambda x: QConfigUnit.from_pytorch(x).as_observer(False))
+
+        torch_version = version.parse(torch.__version__)
+        if (torch_version.major, torch_version.minor) >= (1, 12):
```
Comment (regisss) [inline tests/fx/quantization/test_configuration.py:280]: This is not compatible with the constraint on Torch in Transformers no?

**Review by regisss — APPROVED** [review]
Not super familiar with all this but LGTM! Huge work @michaelbenayoun :rocket:

File: tests/fx/quantization/test_configuration.py (hunk 2 above)
Comment (michaelbenayoun) [reply tests/fx/quantization/test_configuration.py:280]: For now, but could not do otherwise for this specific part.

File: docs/source/_toctree.yml [hunk 3]
```diff
@@ -21,6 +21,8 @@
 - sections:
   - local: fx/optimization
     title: Optimization
+  - local: fx/quantization
```
Comment (fxmarty) [inline docs/source/_toctree.yml:24]: this file is missing

File: optimum/fx/quantization/quantizer.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+def quantize(
+    model: "PreTrainedModel",
+    approach: Union[str, QuantizationApproach],
+    qconfig_dict: Dict[str, Any],
+    preprocess_func: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
+    calibration_dataset: Optional["Dataset"] = None,
+    num_calibration_samples: int = -1,
+    input_names: Optional[List[str]] = None,
+) -> torch.nn.Module:
```
Comment (fxmarty) [inline optimum/fx/quantization/quantizer.py:39]:
```suggestion
) -> torch.fx.GraphModule:
```

File: optimum/fx/quantization/quantizer.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            model.
+        calibration_dataset (`datasets.Dataset`, *optional*):
+            The calibration dataset to use for calibrating the model quantization parameters or QAT.
+        num_calibration_samples (`int`, defaults to -1):
+            The number of examples to use from calibration_dataset.
+        input_names (`List[str]`, *optional*):
+            The name of the inputs to keep when tracing the model for quantization.
+
+    Returns:
+        `torch.nn.Module`:
```
Comment (fxmarty) [inline optimum/fx/quantization/quantizer.py:63]:
```suggestion
        `torch.fx.GraphModule`:
```

File: optimum/fx/quantization/quantizer.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    from datasets import Dataset
+    from torch.ao.quantization.quantize_fx import ObservedGraphModule
+    from transformers import PreTrainedModel
+
+
+def quantize(
+    model: "PreTrainedModel",
+    approach: Union[str, QuantizationApproach],
+    qconfig_dict: Dict[str, Any],
+    preprocess_func: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
```
Comment (fxmarty) [inline optimum/fx/quantization/quantizer.py:35]: There is something wrong with the [ and ]

File: optimum/fx/quantization/quantizer.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        model (`transformers.PreTrainedModel`):
+            The model to quantize.
+        approach (`str` or [`~optimum.utils.runs.QuantizationApproach`]):
+            The quantization approach.
+        qconfig_dict (`Dict[str, Any]`):
+            The dictionary specifying how each part of the model should be quantized.
+            Please refer to [the PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.quantization.quantize_fx.prepare_fx.html#torch.quantization.quantize_fx.prepare_fx)
+            for more details.
+        preprocess_func (`Callable[[Dict[str, Any]], [Dict[str, Any]], *optional*):
+            The preprocessing function to apply to the calibration dataset to make the data ready to be fed to the
```
Comment (fxmarty) [inline optimum/fx/quantization/quantizer.py:53]: We should document / explicit how the `preprocess_func` is called, and have examples in the doc.

File: optimum/fx/quantization/quantizer.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    Returns:
+        `torch.nn.Module`:
+            The quantized version of the model.
+    """
+    prepare_fn = prepare_qat_fx if approach is QuantizationApproach.qat else prepare_fx
+    # TODO: handle the possibility to provide other arguments (equalization qconfig etc)
+    prepared_model = prepare_fn(model, qconfig_dict, input_names=input_names)
+
+    calibration_dataloader = None
+    if calibration_dataset is not None:
```
Comment (fxmarty) [inline optimum/fx/quantization/quantizer.py:71]: To me the logic to preprocess the calibration dataset should be handled by the user, or put in an auxiliary private method.

**Comment by fxmarty** [comment]
It looks nice, more flexible than onnxruntime!

The following code

```python
from transformers import AutoModelForImageClassification

from torch.ao.quantization import get_default_qconfig
from optimum.fx.quantization import prepare_fx, quantize
import torch
from datasets import load_dataset

model_name = "google/vit-base-patch16-224"

model = AutoModelForImageClassification.from_pretrained(model_name)

raw = load_dataset("imagenet-1k", split="train")
raw = raw.select(range(100))

model.eval()
qconfig = get_default_qconfig('fbgemm')
qconfig_dict = {"": qconfig}

quantized_model = quantize(
    model,
    approach="dynamic",
    qconfig_dict=qconfig_dict,
    input_names=["pixel_values"],
)
```

raises a fair bunch of warnings:

```python
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/transformers/utils/fx.py:839: UserWarning: Could not compute metadata for call_function target <function _assert_is_none at 0x7fadb4f4be50>: No metadata was found for Proxy(head_mask_1)
  warnings.warn(f"Could not compute metadata for {kind} target {target}: {e}")
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/transformers/utils/fx.py:839: UserWarning: Could not compute metadata for call_function target <function _assert_is_none at 0x7fadb4f4be50>: No metadata was found for Proxy(labels_1)
  warnings.warn(f"Could not compute metadata for {kind} target {target}: {e}")
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/transformers/utils/fx.py:839: UserWarning: Could not compute metadata for call_function target <function _assert_is_none at 0x7fadb4f4be50>: No metadata was found for Proxy(output_attentions_1)
  warnings.warn(f"Could not compute metadata for {kind} target {target}: {e}")
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/transformers/utils/fx.py:839: UserWarning: Could not compute metadata for call_function target <function _assert_is_none at 0x7fadb4f4be50>: No metadata was found for Proxy(output_hidden_states_1)
  warnings.warn(f"Could not compute metadata for {kind} target {target}: {e}")
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/transformers/utils/fx.py:839: UserWarning: Could not compute metadata for call_function target <function _assert_is_none at 0x7fadb4f4be50>: No metadata was found for Proxy(interpolate_pos_encoding_1)
  warnings.warn(f"Could not compute metadata for {kind} target {target}: {e}")
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/transformers/utils/fx.py:839: UserWarning: Could not compute metadata for call_function target <function _assert_is_none at 0x7fadb4f4be50>: No metadata was found for Proxy(return_dict_1)
  warnings.warn(f"Could not compute metadata for {kind} target {target}: {e}")
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/ao/quantization/observer.py:176: UserWarning: Please use quant_min and quant_max to specify the range for observers.                     reduce_range will be deprecated in a future release of PyTorch.
  warnings.warn(
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/transformers/optimization.py:306: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/ao/quantization/observer.py:1135: UserWarning: must run observer before calling calculate_qparams.                                    Returning default scale and zero point 
  warnings.warn(
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/nn/quantized/_reference/modules/utils.py:25: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  torch.tensor(weight_qparams["scale"], dtype=torch.float, device=device))
/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/nn/quantized/_reference/modules/utils.py:28: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  torch.tensor(weight_qparams["zero_point"], dtype=zero_point_dtype, device=device))
```

Is it expected?

**Comment by fxmarty** [comment]
Also, I could not get the `quantize()` to work for static quantization on vision tasks, related to what appears to me to be a bug in datasets (but I could be doing it wrong): https://github.com/huggingface/datasets/issues/4802

**Comment by fxmarty** [comment]
An other weird behavior I came upon:

```python
from transformers import AutoModelForImageClassification, AutoFeatureExtractor
from transformers.utils.fx import symbolic_trace

from torch.ao.quantization import get_default_qconfig
from optimum.fx.quantization import quantize
from datasets import load_dataset


model_name = "google/vit-base-patch16-224"

model = AutoModelForImageClassification.from_pretrained(model_name)

raw = load_dataset("imagenet-1k", split="validation")
raw = raw.select(range(100))

model.eval()
qconfig = get_default_qconfig('fbgemm')
qconfig_dict = {"": qconfig}

"""
fx_model = quantize(
    model,
    approach="dynamic",
    qconfig_dict=qconfig_dict,
    input_names=["pixel_values"],
)
"""
fx_model = symbolic_trace(model, input_names=["pixel_values"])
preprocessor = AutoFeatureExtractor.from_pretrained(model_name)

for i in range(10):
    inp = preprocessor(raw[i]["image"])
    inp.convert_to_tensors("pt")

    res = fx_model(**inp)

    print(res["logits"].argmax())
```

I get with only doing `symbolic_trace`:
```
tensor(91)
tensor(268)
tensor(979)
tensor(218)
tensor(19)
tensor(658)
tensor(893)
tensor(808)
tensor(305)
tensor(416)
```

while using the commented code with `quantize()` gives:
```
tensor(415)
tensor(415)
tensor(415)
tensor(415)
tensor(415)
tensor(415)
tensor(415)
tensor(415)
tensor(415)
tensor(415)
```

and the logits are always the same.

I use transformers dev version (47e1676255e5dd86b9541f734cd4f4bdcbb50f4a).
